### ISSUE TITLE
design approach

### ISSUE THREAD

**Issue by joeblew99** [murlokswarm/windows#1]
i really like how this is designed to use the webview and inject html via the js bridge.

i have no idea the direction to take with windows. Do you have some design thoughts on this ?

My thoughts:
- c# i think is bad idea. c seams much smarter approach
- then we need a embedded browser based on wekbit or chrome.

**Comment by maxence-charriere** [comment]
Windows in absolutely a pain in the ass! Currently the way I explore is using centennial project and have the web view based on EdgeHTML.
- https://developer.microsoft.com/en-us/windows/bridges/desktop
- https://github.com/Microsoft/DesktopBridgeToUWP-Samples/tree/master/Samples/AppServiceBridgeSample_C%2B%2B

The reason for this is that the rendering is more smooth than webkit or blink and webview is supposed to be used only for graphic rendering.
Also I want the app to be available on the windows store with all the features it provides.

So what is the plan:
- have the go part which uses CGO or loads dll to use **Windows.ApplicationModel.AppService** to communicate with a UWP.
- have a UWP app which embeds a Webview and other stuff like notification, badge, tiles.
- a packager that put these things together via a single command line.

Where I block?
**Windows.ApplicationModel.AppService** is defined with Microsoft not standard C++ syntax.
Was able to create a dll, load it and call this thing but the problem with this approach is how I call Go functions from the dll?

I'm currently exploring the CGO way, i can call standard C++ and Go callback but I did not solve how to call **Windows.ApplicationModel.AppService** from standard C++.

**Comment by maxence-charriere** [comment]
If you want I can do a PR tonight like that you can see where I am.

**Comment by maxence-charriere** [comment]
otherwise there is the Chromium Embeded Framework but honestly, I have been overwhelmed by making this thing work with go :(

**Comment by joeblew99** [comment]
i think its smart to support edge so that we play nice with windows store.
if you want to put up something half finished that is great, to see the direction.

Regarding the Chromium Embedded framework, have you seen gallium ?
https://github.com/alexflint/gallium

**Comment by joeblew99** [comment]
regarding the binding at build time to a dll, does this help ?
https://github.com/golang/go/issues/15894

**Comment by maxence-charriere** [comment]
Thanks for all those links @joeblew99 
Will check that when I have a moment :)

**Comment by influx6** [comment]
@joeblew99 How fast is gallium compared to microsoft EdgeHTML.

**Comment by joeblew99** [comment]
@influx6  gallium uses chromium under the hood. 

But chromium on windows is not what we want. We want to specifically use Edge, so that any apps built get accepted into the Windows Store. Thats what @maxence-charriere  mentioned above and i agree wholeheartedly Windows Store acceptance is a key need.
I fact that Edge has less capabilities should not be too much of an issue because we have the underlying golang runtime to fill in the nasty lack of things in Edge. For example you can use boltDB or QL DB written in golang, and so not rely on Edges implementation of levelDB.

**Comment by maxence-charriere** [comment]
Gallium is based on Chromium so I imagine that Gallium is fast as Chrome.

**Comment by maxence-charriere** [comment]
Guys, look like this more and more close to happen.
![cpptogo](https://cloud.githubusercontent.com/assets/2692731/22622506/ee482abc-eaf0-11e6-9952-6423b33bc6d2.jpg)

**Comment by influx6** [comment]
Nice bro

**Comment by influx6** [comment]
Share the code if you can, would love to hack along

**Comment by rand99** [comment]
How's it going? 
Still on this?

**Comment by maxence-charriere** [comment]
Yeah,
Just pretty busy these times.

**Comment by markwest1** [comment]
Hello:

I'm a professional C#/Java developer with a strong desire to learn Go and introduce it to my development team (a non-trivial challenge). To gain experience with Golang, I have:
* Completed a Pluralsight course
* Done about half the exercises on [exercism.io](http://exercism.io/)'s Golang track
* Implemented the Monkey language interpreter described in Thorsten Ball's [Writing an Interpreter in Go](https://interpreterbook.com/) and described the experience in a [Meetup presentation](https://drive.google.com/open?id=1Udtq9UYeGpkAGKodyv1Qo7xJCPXrXfjtzW-uLjlapBk)
* Partially (~70%) implemented a parser to deserialize [binary-encoded XML (WCF)](https://github.com/carlosfigueira/WCFSamples/tree/master/WCFBinaryTools/WcfBinaryParser)

To help reduce friction from those who might resist adoption of Go, I would like to write a desktop application that will (as a minimum viable product) interact (via REST APIs) with our main product (running in a PaaS) to monitor time-series data and verify correct operation. I'd like the app to run on both Mac and Windows. I've looked at [Gallium](https://github.com/alexflint/gallium) and this project and am trying to decide which to use.

Can you help by describing the pros and cons of each? I don't mind that this project seems to be in its infancy (as well as Gallium), and would be glad to contribute to the application framework, but I need to get it up and running as quickly as possible.

Thanks!

**Comment by maxence-charriere** [comment]
Hello @markwest1, thanks for you interest.
This project is under development and unfortunately the Windows port is not ready. I have something which runs on my machine but it's far to be usable.

Currently only macOS is supported, and there is more work to do to achieve a full operational framework (push notifications and some other thing I did not need yet).

Recently, a lot of people asking about the state of the Windows port. Look like I ll have to push a little more the Windows dev ;).

**Comment by markwest1** [comment]
Thanks for the reply @maxence-charriere ... I just wish I had all my time to write in Go and I would quit my day job :-). When I can, I will install `gcc` and try to build from your `driver-impl` branch.
-Mark

**Comment by maxence-charriere** [comment]
Pros:
- Based on the native browser: Edge for Windows, Safari for macOS. That allow to not embeds the full chromium browser (small exec). Use performance  and battery enhancement coming from Apple or Microsoft optimisations.
- You use only Golang and HTML to build your interface (no Javascript).
- One way to build your app (React like).

Cons:
- Freedom on how to use the framework. No javascript, no webpage, only React like framework.

**Comment by zet4** [comment]
I want to use murlokswarm/app for a project but I need windows (and linux) support, this seems rather dead with last commits many months ago...

Is this still worked on?

**Comment by maxence-charriere** [comment]
Yes, of course. It's just I m working on other branches that doesn't show on the repo metrics.
Currently working on a v2 that make the architecture easier to develop on multiple platform, and use real HTML5 rather than xml.

That project is far to be dead.
- https://github.com/murlokswarm/app/tree/v2
- https://github.com/murlokswarm/app/tree/project-restructuration

It just take time since I m working it on my free time.
